### PR TITLE
AF-2822: Thread starvation while indexing fixed

### PR DESCRIPTION
**Pull request by adrielparedes**
**JIRA**:

[AF-2822](https://issues.redhat.com/browse/AF-2822)

**referenced Pull Requests**:

* https://github.com/kiegroup/appformer/pull/1129
* https://github.com/kiegroup/kie-wb-common/pull/3612

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by ederign** [comment]
@tomasdavidorg if you need help to try out this, please ping Adriel or me and we can share how we tested this (including the problematic repo)

**Comment by adrielparedes** [comment]
Jenkins retest this

**Comment by tomasdavidorg** [comment]
jenkins do fdb

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/test/java/org/uberfire/ext/metadata/io/IOServiceIndexedTest.java [hunk 1]
```diff
@@ -95,7 +97,7 @@ public void setup() throws URISyntaxException {
         when(mockProvider.newFileSystem(any(Path.class), any())).thenReturn(mock(MockFileSystem.class, RETURNS_DEEP_STUBS));
         when(mockProvider.getFileSystem(any(URI.class))).thenReturn(mock(MockFileSystem.class, RETURNS_DEEP_STUBS));
 
-        ioService = spy(new IOServiceIndexedImpl(indexEngine, executorService, indexersFactory, dispatcherFactory));
+        ioService = spy(new IOServiceIndexedImpl(indexEngine, DefaultApplicationScopedProducerutorService, fsWatchExecutorService, indexersFactory, dispatcherFactory));
```
Comment (tomasdavidorg) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/test/java/org/uberfire/ext/metadata/io/IOServiceIndexedTest.java:100]: This caused the FDB failed.
```suggestion
        ioService = spy(new IOServiceIndexedImpl(indexEngine, indexingExecutorService, fsWatchExecutorService, indexersFactory, dispatcherFactory));
```

**Comment by tomasdavidorg** [comment]
@adrielparedes It looks like the tests are not executed in `uberfire-metadata-commons-io` module https://github.com/kiegroup/appformer/blob/master/uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/pom.xml#L210 The `<include>` should contain `<include>**/*Test.java</include>`.

**Comment by tomasdavidorg** [comment]
Hi  @adrielparedes,

I did some testing on my local machine (CPU 1.90GHzx8,32GB RAM) and I compared indexing in these two builds:
- RHPAM 7.10.1 (**PAM**)
- WAR from this PR (**FIX**)
I started both on EAP 7.3.5 with params `-Xms2G -Xmx4G -XX:MetaspaceSize=1G -XX:MaxMetaspaceSize=2G`.

For the fix I used the settings you sent me:
- `JAVA_OPTS="$JATA_OPTS -Dorg.appformer.concurrent.indexing.thread.limit=9"`
- `<singleton default-access-timeout="60000"/>`

#### Results:
|Action|1|2|3|4|5|Avarage|Result|
|-|-|-|-|-|-|-|-|
|Import Mortgage process (PAM)|6s|6s|7s|7s|7s|6.6s|Original|
|Import Mortgage process (FIX)|6s|6s|6s|7s|7s|6.4s|-0.2s (-3.03 %)|
|Import 100 assets project - 1 branch (PAM)|22s|18s|19s|20s|20s|19.8s|Original|
|Import 100 assets project - 1 branch (FIX)|20s|18s|20s|19s|19s|19.2s|-0.8s (-3.03 %)|
|Import 100 assets project - 10 branches (PAM)|85s|91s|105s|115s|96s|98.4s|Original|
|Import 100 assets project - 10 branches (FIX)|206s|197s|197s|196s|205s|200.2s|+101.8s (+103.45 %)|
|Import 100 assets project - 100 branches (PAM)|-|-|-|-|-|28m|Original|
|Import 100 assets project - 100 branches (FIX)|-|-|-|-|-|32m|+4m (+14.28 %)|
|Index .niogit 10 projects with 100 assets - 100 branches (PAM)|-|-|-|-|-|Out of memory|Didn't finish|
|Index .niogit 10 projects with 100 assets - 100 branches (FIX)|-|-|-|-|-|3h 17min|Finished|

**Comment by adrielparedes** [comment]
jenkins retest this

**Comment by adrielparedes** [comment]
jenkins run fdb

**Comment by adrielparedes** [comment]
jenkins execute fdb

**Review by tomasdavidorg — APPROVED** [review]

**Comment by adrielparedes** [comment]
@tomasdavidorg you are right, I forgot to fix the test, sorry!

**Comment by adrielparedes** [comment]
jenkins run fdb

**Review by ederign — APPROVED** [review]

**Comment by ederign** [comment]
Awesome work @adrielparedes and @tomasdavidorg thanks for testing it!

**Pull request merged**